### PR TITLE
Various improvements

### PR DESCRIPTION
**Pull request by sebv**
Various improvement:
- package upgrade
- use chai
- use B
- use equivalent B method when they exist instead of custom code.
- adds waitForCondition
- fixed some test flakes

### CODE REVIEW

File: lib/asyncbox.js [hunk 1]
```diff
@@ -43,34 +41,11 @@ async function retryInterval (times, sleepMs, fn, ...args) {
...
-      results.push(res);
-      if (results.length === promises.length) {
-        done.resolve(results);
-      }
-    }, function (err) {
-      done.reject(err);
-    });
-  }
-  return done.promise;
+  return await B.all(promises);
```
Comment (jlipps) [inline lib/asyncbox.js]: `B.all` has exactly the same behavior, eh?

File: lib/asyncbox.js [hunk 2]
```diff
@@ -43,34 +41,11 @@ async function retryInterval (times, sleepMs, fn, ...args) {
...
 }
 
 function nodeify (promisey, cb) {
-  promisey.then(res => { cb(null, res); }, cb);
+  return B.resolve(promisey).nodeify(cb);
```
Comment (jlipps) [inline lib/asyncbox.js]: this also handles the error case?

Comment (sebv) [inline lib/asyncbox.js]: yes

File: lib/asyncbox.js [hunk 3]
```diff
@@ -87,9 +62,7 @@ function nodeifyAll (promiseyMap) {
 }
 
 function asyncify (fn, ...args) {
-  fn.call(null, ...args).then(() => {}, (err) => {
-    console.error(err.stack);
-  });
+  B.resolve(fn.call(null, ...args)).done();
```
Comment (jlipps) [inline lib/asyncbox.js]: `done` takes care of printing the error? or it throws it in the calling context?

**Comment by sebv** [comment]
cleaned up commit order.

File: lib/asyncbox.js [hunk 4]
```diff
@@ -123,5 +96,30 @@ async function asyncfilter (coll, filter, runInParallel = true) {
...
+async function waitForCondition (condFn, opts = {}) {
+  _.defaults(opts, {
+    waitMs: 5000,
+    intervalMs : 500,
+  });
+  let debug = opts.logger ? opts.logger.debug.bind(opts.logger) : () => {};
+  let begunAt = Date.now();
+  let endAt = begunAt + opts.waitMs;
+  let spin = async () => {
+    if(await condFn()) {
```
Comment (jlipps) [inline lib/asyncbox.js]: `if (`

File: lib/asyncbox.js [hunk 5]
```diff
@@ -123,5 +96,30 @@ async function asyncfilter (coll, filter, runInParallel = true) {
...
+      return;
+    }
+    let now = Date.now();
+    let waited = now - begunAt;
+    if (now < endAt) {
+      debug("Waited for " + waited + "ms so far");
```
Comment (jlipps) [inline lib/asyncbox.js]: interpolated string

File: lib/asyncbox.js [hunk 6]
```diff
@@ -123,5 +96,30 @@ async function asyncfilter (coll, filter, runInParallel = true) {
...
+      await B.delay(opts.intervalMs);
+      await spin();
+    } else {
+      throw new Error("Condition unmet after " + waited + "ms. Timing out.");
```
Comment (jlipps) [inline lib/asyncbox.js]: interpolated string

**Comment by jlipps** [comment]
ok made some small comments. can you also add docs for waitForCondition in the readme, or at least a TODO that we can fill out later

File: lib/asyncbox.js (hunk 1 above)
Comment (sebv) [inline lib/asyncbox.js]: https://github.com/petkaantonov/bluebird/blob/master/API.md#all---promise
Looks like it. I don't think the Q one has.

btw parallel is a bad name, the promises have already started. Usually control flow methods take function returning promises as arguments.

File: lib/asyncbox.js (hunk 3 above)
Comment (sebv) [inline lib/asyncbox.js]: Just reports the error, I don't think it is possible to throw in the calling context, there are cases where it doesn't exists.

Comment (jlipps) [inline lib/asyncbox.js]: ok. just wanted to make sure it at least logs it.

**Comment by sebv** [comment]
Fixes done, added doc.

File: lib/asyncbox.js (hunk 3 above)
Comment (sebv) [inline lib/asyncbox.js]: just noticed, no need for `call`

Comment (jlipps) [inline lib/asyncbox.js]: yeah why not just `fn(...args)`

**Comment by jlipps** [comment]
@sebv published as 2.3.0

**Pull request merged**